Sema: Check the availability of conformance type witnesses

tshortli opened the pull request:
The type satisfying a protocol requirement must be at least as available as the associated type for the requirement. This has always been a potential soundness hole that should have been checked, but it is especially important to diagnose these issues now that associated type requirements can have their own availability (https://github.com/swiftlang/swift/pull/70735). Unfortunately, these diagnostics cannot be introduced as errors since doing so would break source compatibility for existing libraries so for now they are staged as warnings until a post Swift 6 language mode is available.

Resolves rdar://134093006.

tshortli commented:
@swift-ci please test

tshortli commented:
@swift-ci please test source compatibility

slavapestov commented on test/decl/protocol/associated_type_witness_availability.swift:
```diff
@@ -0,0 +1,83 @@
+// RUN: %swift -typecheck -verify -target %target-cpu-apple-macosx12 %s
+// REQUIRES: OS=macosx
+
+protocol ProtoWithAssocType {
```
Can you add a test where the witness is inferred, just in case?

```
protocol ProtoWithAssocType {
  associatedtype A
  func f(_: A)
}


@available(...) struct SomeType {}

struct S: ProtoWithAssocType {
  func f(_: SomeType)
}

tshortli replied:
Good catch, inference will need to diagnose when the availability of the inferred type is insufficient I guess.

tshortli replied:
If these availability checks had been around from the beginning I suppose it's possible that they would have been implemented in associated type inference to reject the (potentially) unavailable candidates and look for others. But I think I'm going to keep these checks after associated type inference for now because it's going to be a lot to move them into inference. I've also discovered that there are source compatibility issues with existing frameworks, so the diagnostics need to be downgraded to warnings in existing language modes, which would be extra messy to do in type inference. To successfully diagnose the inference cases I just need to look through the inferred type to its original declaration, instead of checking the availability of the direct witness decl, which may be a synthesized typealias.

tshortli commented:
@swift-ci please smoke test

tshortli committed:
AST: Take decls instead of identifiers in some conformance diagnostics.

NFC.

tshortli committed:
Sema: Expose a utility on ExportContext for checking unavailability.

Code that wants to determine whether a given declaration should be considered
unavailable should use this new utility, which will take the unavailability of
the context of use into consideration.

NFC.

tshortli committed:
Sema: Check availability of conformance type witnesses.

The type satisfying a protocol requirement must be at least as available as the
associated type for the requirement.

Resolves rdar://134093006

tshortli commented:
@swift-ci please smoke test
